crates/sui-graphql-rpc/src/types/move_package.rs

```diff
@@ -81,96 +82,63 @@ impl MovePackage {
-                        self.native_object.id()
-                    ),
-                )
-                .into());
-            }
-
-            let mut forward = true;
-            let mut count = first.unwrap_or(DEFAULT_PAGE_SIZE as u64);
-            count = last.unwrap_or(count);
-
-            let mut mod_list = mod_map
-                .clone()
-                .into_iter()
-                .collect::<Vec<(String, Vec<u8>)>>();
-
-            // ok to unwrap because we know mod_map is not empty
-            let mut start = &if last.is_some() {
-                forward = false;
-                mod_list.last().map(|c| c.0.clone())
-            } else {
-                mod_list.first().map(|c| c.0.clone())
-            }
-            .unwrap();
-
-            if let Some(aft) = &after {
-                start = aft;
-            } else if let Some(bef) = &before {
-                start = bef;
-                forward = false;
-            };
-
-            if !forward {
-                mod_list = mod_list.into_iter().rev().collect();
-            }
-
-            let mut res: Vec<_> = mod_list
-                .iter()
-                .skip_while(|(name, _)| name.as_str() != start)
-                .skip((after.is_some() || before.is_some()) as usize)
-                .take(count as usize)
-                .map(|(name, module)| {
-                    CompiledModule::deserialize_with_config(
-                        module,
-                        move_binary_format::file_format_common::VERSION_MAX,
-                        true,
-                    )
-                    .map(|x| (name, MoveModule { native_module: x }))
-                })
-                .collect::<Result<Vec<_>, _>>()?;
-            let mut has_prev_page = &mod_list.first().unwrap().0 != start;
-            let mut has_next_page =
-                !res.is_empty() && &mod_list.last().unwrap().0 != res.last().unwrap().0;
-            if !forward {
-                res = res.into_iter().rev().collect();
-                has_prev_page = &mod_list.first().unwrap().0 != start;
-                has_next_page =
-                    !res.is_empty() && &mod_list.last().unwrap().0 != res.first().unwrap().0;
-            }
-
-            let mut connection = Connection::new(has_prev_page, has_next_page);
-
-            connection.edges.extend(
-                res.into_iter()
-                    .map(|(name, module)| Edge::new(name.clone(), module)),
-            );
-            return Ok(Some(connection));
+        let parsed = self.parsed_package()?;
+        let module_range = parsed.modules().range((
+            after.map_or(B::Unbounded, B::Excluded),
+            before.map_or(B::Unbounded, B::Excluded),
+        ));
+
+        let total = module_range.clone().count() as u64;
+        let (skip, take) = match (first, last) {
+            (Some(first), Some(last)) if last < first => (first - last, last),
```

Comment:
Maybe my reading comprehension skills are just poor lol, but in the pr desc it sounds like we select the first 'first', select the last 'last', and then join them together. looking at the implementation, it looks like we slice `first_slice = edges[:first]` and then `final_slice=first_slice[last:]`. I think the latter/ in the code is the correct implementation for graphql cursor specification, so just double checking

Reply:
Your code comprehension is correct, and that's what I was trying to convey in English (and how I understood the spec as well), I was just doing it poorly!